src/main/tls.c

```diff
@@ -3897,7 +3897,7 @@ fr_tls_status_t tls_ack_handler(tls_session_t *ssn, REQUEST *request)
 		return FR_TLS_FAIL;
 
 	case handshake:
-		if ((ssn->info.handshake_type == handshake_finished) && (ssn->dirty_out.used == 0)) {
+		if (SSL_is_init_finished(ssn->ssl) && (ssn->dirty_out.used == 0)) {
```

Comment:
That should likely be `tls_session->is_init_finished`, which was added for 3.0.18

Reply:
OK - if so should d90c4bf807c66 in master also be changed?

Reply:
Yes, likely

Reply:
Why? It’s better to get the information directly from the SSL *?

Reply:
Because the change to add `tls_session->is_init_finished` wasn't done on a whim.  See:

https://github.com/FreeRADIUS/freeradius-server/blob/v3.0.x/src/main/tls.c#L792

While the change is perhaps not needed here, it's best to have one way of doing things.  If we randomly do API calls vs checking local variables, then we have a higher potential of doing the wrong thing.

Reply:
is_init_finished is only used to implement an edge based trigger to add the TLS-Session-Version attribute to the session list. The master branch code works differently and doesn't require a hack to implement the edge based trigger.

Reply:
`is_init_finished` is used for more than that in v3.  And this patch set is for v3.

Reply:
My main concern is code from v3.0.x being applied to the master branch without careful consideration.

Reply:
My issue was with @mcnewton asking if the master branch should be changed and your response of:

> Yes, likely

That's why I was stating that it wasn't required in master.  I agree in v3.0.x that we should use the shadowed value.

Reply:
So agreed `ssn->is_init_finished` in v3, and leave master as-is (using `SSL_is_init_finished()`)?

Reply:
That's fine.